master/buildbot/www/hooks/bitbucket.py

```diff
@@ -12,24 +12,170 @@
 
 from twisted.python import log
 
+try:
+    import json
+    assert json
+except ImportError:
+    import simplejson as json
 
-def getChanges(request, options=None):
-    """Catch a POST request from BitBucket and start a build process
+
+_HEADER_CT = 'Content-Type'
+_HEADER_EVENT = 'X-Event-Key'
+
+
+class BitbucketEventHandler(object):
+
+    def __init__(self, codebase=None):
+        self._codebase = codebase
+
+    def process(self, request):
+        """
+        Process a webhook event request.
+
+        See example on https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/hooks
+        for list of all the different events.
+        """
+        payload = self._get_payload(request)
+        event_type = request.getHeader(_HEADER_EVENT)
+        log.msg("Processing event %s: %r" % (_HEADER_EVENT, event_type,), logLevel=logging.DEBUG)
+        event_type = event_type.replace(":", "_")
+        handler = getattr(self, 'handle_%s' % event_type, None)
+
+        if handler is None:
+            raise ValueError('Unknown event: %r' % (event_type,))
+
+        return handler(payload)
+
+    def _get_payload(self, request):
+        content = request.content.read()
+        content_type = request.getHeader(_HEADER_CT)
+
+        if content_type == 'application/json':
+            payload = json.loads(content)
+        elif content_type == 'application/x-www-form-urlencoded':
+            payload = json.loads(request.args['payload'][0])
+        else:
+            raise ValueError('Unknown content type: %r' % (content_type,))
+
+        log.msg("Payload: %r" % payload, logLevel=logging.DEBUG)
+
+        return payload
+
+    def handle_repo_push(self, payload):
+        repo = payload['repository']
+        project = repo['name']
+        push = payload['push']
+        payload['ref'] = push['changes'][0]["new"]["name"]
+        repo_url = "git@bitbucket.org:%s.git" % payload['repository']['full_name']  # TOFIX: Hard coded as clone URL is not available
+        changes = self._process_change(payload, repo, repo_url, project)
+        log.msg("Received %d changes from Bitbucket" % len(changes))
+        return changes, 'git'
+
+    def handle_pullrequest_updated(self, payload):
+        return self.handle_pullrequest(payload, 'updated_on')
+
+    def handle_pullrequest_created(self, payload):
+        return self.handle_pullrequest(payload, 'created_on')
+
+    def _process_change(self, payload, repo, repo_url, project):
+        """
+        Consumes the JSON as a python object and returns the changes.
+
+        :arguments:
+            payload
+                Python Object that represents the JSON sent by Bitbucket Webhook event.
+        """
+        changes = []
+
+        #  TOFIX: Code from github hook code, is this functionality needed?
+        # refname = payload['ref']
+        #
+        #  We only care about regular heads, i.e. branches
```

Comment:
I think you can remove that code. This functionality should be achieved using event filters.

The gerrit change source has a filter in the changesource directly because gerrit canhost lots of repositories, so there is an advantage to filter at the source to not flood the db with ininterresting events.

I am not sure if this can apply there